S7.Net/PLC.cs

```diff
@@ -107,11 +107,35 @@ public bool IsConnected
             {
                 try
                 {
+                    bool clientConnected = false;
+
                     if (tcpClient == null)
                         return false;
 
                     //TODO: Actually check communication by sending an empty TPDU
-                    return tcpClient.Connected;
+                    if (tcpClient.Client.Poll(0,SelectMode.SelectRead))
+                    {
+                        byte[] buffer = new byte[1];                        
+                        if (tcpClient.Client.Receive(buffer, SocketFlags.Peek) == 0)
```

Comment:
The receive method will block as long as it needs to read data, it only immediately returns when the socket has been closed. My (personal and professional) opinion: don't use the sync `Receive`, but use `ReceiveAsync` (if you need to use either that is). From this it follows that you can't use it in a property, which is fine, because you can't assert that communication will be successful anyway. Also, I thought I actually copied the comment from `TcpClient.IsConnected` to this `IsConnected` property (but apparently I didn't?), which clearly states that you can use it to see if the last communication attempt has been succesful and that you can't really rely on it anyway. I'd like to keep the IsConnected property as simple as that.

All this doesn't mean there's no room for improvement, but perhaps it would just be a good idea to fire a read for DB1,1 and see if a valid TPKT is returned (which could have an error indication in the message, but will still tell if the connection is fine).